intake_esm/core.py

```diff
@@ -707,3 +715,9 @@ def _make_entry(key, df, aggregation_info):
         name=key, description='', driver='esm_group', args=args, metadata={}
     )
     return entry
+
+
+def _get_groups_and_keys(df, aggregation_info):
```

Comment:
since this is only used inside the catalog class, maybe make this a class method. Something like:

```python
def _set_groups_and_keys(self):
    self._grouped = self._df.groupby(aggregation_info['groupby_attrs'])
    self._keys = list(grouped.groups.keys())
```

Then above, you would just call the method after setting `_df`:

```python
self._set_groups_and_keys()
```

Reply:
Done!